Please fiix 

0s
12s
Run ludeeus/action-shellcheck@00cae500b08a931fb5698e11e79bfbd38e612a38
Run if [[ "Linux" == "macOS" ]]; then
tar: This does not look like a tar archive
xz: (stdin): File format not recognized
tar: Child returned status 1
tar: Error is not recoverable: exiting now
Error: Process completed with exit code 2.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -234,7 +234,7 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 10
     # Upstream ShellCheck project: https://github.com/koalaman/shellcheck
-    # Action wrapper used below: ludeeus/action-shellcheck (runs koalaman/shellcheck).
+    # Invoked below via the `shellcheck` binary pre-installed on ubuntu-latest.
 
     steps:
       - name: Checkout code

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -269,16 +269,28 @@ jobs:
           done < <(find . -name "*.sh" -not -path "./target/*" -not -path "./.git/*")
           exit $EXIT_CODE
 
-      - name: Run ShellCheck (koalaman/shellcheck via ludeeus/action-shellcheck)
-        # Pinned to a tagged release (Issue #24) — `@master` is unpinned and
-        # a compromised upstream commit would silently alter CI behaviour.
-        uses: ludeeus/action-shellcheck@00cae500b08a931fb5698e11e79bfbd38e612a38  # v2.0.0
-        with:
-          scandir: "."
-          severity: warning
-          ignore_paths: "target .git"
+      - name: Run ShellCheck (koalaman/shellcheck, pre-installed on the runner)
+        # Invoke the ShellCheck binary pre-installed on ubuntu-latest rather
+        # than downloading a release asset at job time. The previous
+        # ludeeus/action-shellcheck wrapper fetched the binary with
+        # `curl -Lso` (no `-f`), so a transient release-asset download
+        # error saved an HTML error page and the tar/xz extraction failed
+        # with "does not look like a tar archive" (exit 2). Using the
+        # pre-installed binary removes that network dependency and also
+        # drops a third-party action from the supply chain.
         env:
           SHELLCHECK_OPTS: -s bash
+        run: |
+          shellcheck --version
+          scripts=()
+          while IFS= read -r script; do
+            scripts+=("$script")
+          done < <(find . -name "*.sh" -not -path "./target/*" -not -path "./.git/*")
+          if [ ${#scripts[@]} -eq 0 ]; then
+            echo "No shell scripts found — nothing to lint"
+            exit 0
+          fi
+          shellcheck --severity=warning "${scripts[@]}"
 
       - name: Install bats
         run: sudo apt-get update && sudo apt-get install -y bats

--- a/scripts/check-shellcheck-dedup.sh
+++ b/scripts/check-shellcheck-dedup.sh
@@ -10,8 +10,12 @@
 # aggregator that branch protection gates on.
 #
 # This guard asserts that invariant: exactly one workflow under
-# `.github/workflows` may invoke `ludeeus/action-shellcheck`. Re-introducing a
-# second invocation re-creates the duplicated maintenance surface and fails.
+# `.github/workflows` may invoke ShellCheck. ShellCheck can be invoked either
+# via the `ludeeus/action-shellcheck` action or by calling the pre-installed
+# `shellcheck` binary directly in a `run:` step (PR #184 switched `ci.yml` to
+# the latter to drop the flaky release-asset download). Re-introducing a second
+# invocation by either method re-creates the duplicated maintenance surface and
+# fails.
 #
 # The script takes an optional `--workflows DIR` argument so BATS tests can
 # exercise it against fixtures. With no argument it scans
@@ -27,9 +31,10 @@ Options:
                     .github/workflows relative to the repo root).
   -h, --help        Show this message.
 
-Exits 0 when exactly one workflow invokes ludeeus/action-shellcheck. Exits
-non-zero with a descriptive message when ShellCheck is missing entirely or
-duplicated across more than one workflow file.
+Exits 0 when exactly one workflow invokes ShellCheck (via
+ludeeus/action-shellcheck or a direct `shellcheck --severity` run step).
+Exits non-zero with a descriptive message when ShellCheck is missing entirely
+or duplicated across more than one workflow file.
 EOF
 }
 
@@ -62,13 +67,16 @@ if [[ ! -d "$WORKFLOWS_DIR" ]]; then
   exit 2
 fi
 
-# Collect every workflow file that *invokes* ShellCheck. We match `uses:` lines
-# only (ignoring leading `- `) and skip comment lines so that prose mentioning
-# the action does not count as an invocation.
+# Collect every workflow file that *invokes* ShellCheck. A workflow counts when
+# it either references the action on a `uses:` line (ignoring leading `- `) or
+# runs the `shellcheck --severity` binary in a `run:` step. Both patterns anchor
+# to the start of the line so prose mentioning ShellCheck in a comment does not
+# count as an invocation.
 matches=()
 while IFS= read -r workflow; do
   [[ -f "$workflow" ]] || continue
-  if grep -qE '^[[:space:]]*(-[[:space:]]*)?uses:[[:space:]]*ludeeus/action-shellcheck@' "$workflow"; then
+  if grep -qE '^[[:space:]]*(-[[:space:]]*)?uses:[[:space:]]*ludeeus/action-shellcheck@' "$workflow" \
+    || grep -qE '^[[:space:]]*shellcheck[[:space:]]+--severity' "$workflow"; then
     matches+=("$workflow")
   fi
 done < <(find "$WORKFLOWS_DIR" -maxdepth 1 \( -name "*.yml" -o -name "*.yaml" \) -type f | sort)
@@ -81,7 +89,7 @@ if [[ "$count" -eq 1 ]]; then
 fi
 
 if [[ "$count" -eq 0 ]]; then
-  echo "FAIL no workflow invokes ludeeus/action-shellcheck — ShellCheck coverage is missing." >&2
+  echo "FAIL no workflow invokes ShellCheck — ShellCheck coverage is missing." >&2
   exit 1
 fi
 

--- a/tests/scripts/shellcheck_dedup.bats
+++ b/tests/scripts/shellcheck_dedup.bats
@@ -38,6 +38,26 @@ jobs:
 EOF
 }
 
+# Write a workflow that invokes the pre-installed shellcheck binary directly
+# in a run step (the form ci.yml uses after PR #184 dropped the action).
+write_shellcheck_run_workflow() {
+  local file="$1"
+  cat >"$file" <<'EOF'
+name: Example Run
+on:
+  pull_request:
+jobs:
+  shellcheck:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v5
+      - name: Run ShellCheck (pre-installed)
+        run: |
+          shellcheck --version
+          shellcheck --severity=warning script.sh
+EOF
+}
+
 # Write a workflow that does NOT invoke ShellCheck (only mentions it in prose).
 write_unrelated_workflow() {
   local file="$1"
@@ -61,6 +81,23 @@ EOF
   [ "$status" -eq 0 ]
   [[ "$output" == *"exactly one workflow"* ]]
   [[ "$output" == *"ci.yml"* ]]
+}
+
+@test "counts a direct shellcheck run step as an invocation" {
+  write_shellcheck_run_workflow "$TMP_WF/ci.yml"
+  write_unrelated_workflow "$TMP_WF/build.yml"
+  run "$SCRIPT_UNDER_TEST" --workflows "$TMP_WF"
+  [ "$status" -eq 0 ]
+  [[ "$output" == *"exactly one workflow"* ]]
+  [[ "$output" == *"ci.yml"* ]]
+}
+
+@test "fails when a run step duplicates the action invocation" {
+  write_shellcheck_workflow "$TMP_WF/ci.yml"
+  write_shellcheck_run_workflow "$TMP_WF/shellcheck.yml"
+  run "$SCRIPT_UNDER_TEST" --workflows "$TMP_WF"
+  [ "$status" -ne 0 ]
+  [[ "$output" == *"duplicated across 2 workflows"* ]]
 }
 
 @test "fails when ShellCheck is duplicated across two workflows" {

